lib/block_view.go

```diff
@@ -7620,6 +7911,166 @@ func (bav *UtxoView) _connectSwapIdentity(
+	if len(ownerPublicKey) != btcec.PubKeyBytesLenCompressed {
+		return 0, 0, nil, RuleErrorAuthorizeDerivedKeyInvalidOwnerPublicKey
+	}
+	if _, err := btcec.ParsePubKey(ownerPublicKey, btcec.S256()); err != nil {
+		return 0, 0, nil, errors.Wrap(RuleErrorAuthorizeDerivedKeyInvalidOwnerPublicKey, err.Error())
+	}
+
+	// Validate the derived public key.
+	derivedPublicKey := txMeta.DerivedPublicKey
+	if len(derivedPublicKey) != btcec.PubKeyBytesLenCompressed {
+		return 0, 0, nil, RuleErrorAuthorizeDerivedKeyInvalidDerivedPublicKey
+	}
+	if _, err := btcec.ParsePubKey(derivedPublicKey, btcec.S256()); err != nil {
+		return 0, 0, nil, errors.Wrap(RuleErrorAuthorizeDerivedKeyInvalidDerivedPublicKey, err.Error())
+	}
+
+	// Verify that the access signature is valid. This means the derived key is authorized.
+	err := _verifyAccessSignature(ownerPublicKey, derivedPublicKey,
+		txMeta.ExpirationBlock, txMeta.AccessSignature)
+	if err != nil {
+		return 0, 0, nil, errors.Wrap(RuleErrorAuthorizeDerivedKeyAccessSignatureNotValid, err.Error())
+	}
+
+	// Get current (previous) derived key entry. We might revert to it later so we copy it.
+	prevDerivedKeyEntry := bav._getDerivedKeyMappingForOwner(ownerPublicKey, derivedPublicKey)
+
+	// Authorize transactions can be signed by both owner and derived keys. However, this
+	// poses a risk in a situation where a malicious derived key, which has previously been
+	// de-authorized by the owner, were to attempt to re-authorize itself.
+	// To prevent this, the following check completely blocks a derived key once it has been
+	// de-authorized. This makes the lifecycle of a derived key more controllable.
+	if prevDerivedKeyEntry != nil && !prevDerivedKeyEntry.isDeleted {
+		if prevDerivedKeyEntry.OperationType == AuthorizeDerivedKeyOperationNotValid {
+			return 0, 0, nil, RuleErrorAuthorizeDerivedKeyDeletedDerivedPublicKey
+		}
+	}
+
+	// At this point we've verified the access signature, which means the derived key is authorized
+	// to sign on behalf of the owner. In particular, if this authorize transaction was signed
+	// by the derived key, we would accept it. We accommodate this by adding a temporary derived
+	// key entry to UtxoView, to support first-time derived keys (they don't exist in the DB yet).
+	// As a result, and if the derived key is present in transaction's ExtraData, we will
+	// pass signature verification in _connectBasicTransfer() -> _verifySignature().
+	//
+	// NOTE: Setting a mapping in UtxoView prior to fully validating a transaction shouldn't be
+	// reproduced elsewhere. It's error-prone, controversial, some even call it "a dirty hack!"
+	// All considered, this feature greatly simplifies the flow in identity - from the moment you
+	// generate a derived key, you can use it to sign any transaction offline, including authorize
+	// transactions. It also resolves issues in situations where the owner account has insufficient
+	// balance to submit an authorize transaction.
```

Comment:
Awesome comment. Always tend toward over-commenting like this!

Reply:
I wonder whose quote is it? 🧐